opt: panic: rowCount passed in was too small

See #34440.

```
CREATE TABLE t (i INT8, f FLOAT8, d DECIMAL, s STRING, b BYTES, z BOOL);

WITH
    with_46
        (
            col_484,
            col_485,
            col_486,
            col_487,
            col_488,
            col_489,
            col_490,
            col_491,
            col_492,
            col_493,
            col_494,
            col_495,
            col_496,
            col_497,
            col_498,
            col_499,
            col_500
        )
        AS (
            SELECT
                *
            FROM
                (
                    VALUES
                        (
                            CASE
                            WHEN CASE WHEN true THEN false ELSE false END THEN '6497:2675:4b8d:3c60:8dce:4d81:50e9:ca16/124'
                            ELSE '89d0:3cc:9f49:e53f:819a:e42f:7ffb:4a5f/92' - 5477552741327549989
                            END
                            | CASE
                                WHEN not_like_escape(e'_h\\x0b1\\x01^', e'XQ\\b}fx\\fY\\x0f', 'aj_')
                                THEN COALESCE('153.175.79.216':::INET, inet_server_port() + NULL)
                                ELSE 'f546:fe2e:e963:b1f1:b5a1:a379:49fd:2736/30':::INET
                                END,
                            false,
                            B'1001111110001000011110111010110100010110111011000000101011010000000101001110100001010001',
                            2458092858580231762:::INT8,
                            'c955f141-6988-4942-aa99-95455e6c7c3d':::UUID,
                            1494121970:::OID,
                            '1970-01-09 06:42:02.000575+00:00':::TIMESTAMPTZ,
                            NULL,
                            B'11101010001111010001110010110110100001101110000001010100101001110010011010010001100011100100',
                            '1978-08-08':::DATE,
                            '15:04:01.556338':::TIME,
                            '1970-01-05 10:15:02.0007+00:00':::TIMESTAMP,
                            7.407954582851022096E+22:::DECIMAL,
                            NULL,
                            1417320083:::OID,
                            1.8336144724179084:::FLOAT8,
                            NULL
                        )
                )
                    AS tab_107
                        (
                            col_484,
                            col_485,
                            col_486,
                            col_487,
                            col_488,
                            col_489,
                            col_490,
                            col_491,
                            col_492,
                            col_493,
                            col_494,
                            col_495,
                            col_496,
                            col_497,
                            col_498,
                            col_499,
                            col_500
                        )
            EXCEPT
                SELECT
                    *
                FROM
                    (
                        VALUES
                            (
                                '211.133.117.16/3':::INET,
                                false,
                                B'111000010011100011101111000000101001010000000100111010101111011',
                                (-4027224010634193633):::INT8,
                                'cc4da225-be8e-44b8-a36f-da90b97ed83f':::UUID,
                                NULL,
                                NULL,
                                e'\\\\xa6e9d0e8639a519557':::BYTES,
                                B'0110011101101110100111011111000101101010010100101110001011001100',
                                '1990-12-02':::DATE,
                                '20:31:18.06948':::TIME,
                                '1970-01-05 19:45:20.000883+00:00':::TIMESTAMP,
                                0.06826736946325525687:::DECIMAL,
                                '8f672d71-fa6d-4bfe-b3c9-d5ba51773aee':::UUID,
                                716023061:::OID,
                                0.1461138164507641:::FLOAT8,
                                '03:28:13.179218':::TIME
                            )
                    )
                        AS tab_108
                            (
                                col_501,
                                col_502,
                                col_503,
                                col_504,
                                col_505,
                                col_506,
                                col_507,
                                col_508,
                                col_509,
                                col_510,
                                col_511,
                                col_512,
                                col_513,
                                col_514,
                                col_515,
                                col_516,
                                col_517
                            )
                INTERSECT
                    SELECT
                        *
                    FROM
                        (
                            VALUES
                                (
                                    '121.98.217.25/25':::INET,
                                    true,
                                    B'0001101010110100101000100110011',
                                    6230263107624348070:::INT8,
                                    'd0fedc79-631b-4429-95f0-e53e7056f181':::UUID,
                                    1310614926:::OID,
                                    '1970-01-09 23:21:56.000117+00:00':::TIMESTAMPTZ,
                                    NULL,
                                    B'110000111001001010011011100101100111001',
                                    '1987-09-03':::DATE,
                                    NULL,
                                    '1970-01-01 07:36:37.000835+00:00':::TIMESTAMP,
                                    35419685913.7819160:::DECIMAL,
                                    '1a47ad56-1d2b-433c-a276-1acd151f96c4':::UUID,
                                    3640439885:::OID,
                                    0.0954438200827454:::FLOAT8,
                                    '23:19:54.326524':::TIME
                                )
                        )
                            AS tab_109
                                (
                                    col_518,
                                    col_519,
                                    col_520,
                                    col_521,
                                    col_522,
                                    col_523,
                                    col_524,
                                    col_525,
                                    col_526,
                                    col_527,
                                    col_528,
                                    col_529,
                                    col_530,
                                    col_531,
                                    col_532,
                                    col_533,
                                    col_534
                                )
            EXCEPT
                SELECT
                    *
                FROM
                    (
                        VALUES
                            (
                                'e4f1:b3a5:52ad:c731:ad14:42b5:5df7:eaa6/85':::INET,
                                NULL,
                                NULL,
                                6855942593912848994:::INT8,
                                'd64c6605-91d9-4a05-9bb0-ada5d7267412':::UUID,
                                2982045989:::OID,
                                '1970-01-04 21:47:21.000668+00:00':::TIMESTAMPTZ,
                                e'\\\\x7fe4f6e206d9271def':::BYTES,
                                B'0111101110001110010110011101010100111010100111010001010110010001001001101',
                                '1987-02-26':::DATE,
                                NULL,
                                '1970-01-12 09:34:12.000613+00:00':::TIMESTAMP,
                                4.328275068256031932E+19:::DECIMAL,
                                'da4ec56c-5c49-434d-98ed-690087fee5b8':::UUID,
                                609937469:::OID,
                                1.991322968928448:::FLOAT8,
                                NULL
                            ),
                            (
                                '99.18.28.134/1':::INET,
                                true,
                                NULL,
                                (-5845805165557115523):::INT8,
                                NULL,
                                3482780434:::OID,
                                '1970-01-02 22:22:25.000758+00:00':::TIMESTAMPTZ,
                                e'\\\\x6a':::BYTES,
                                B'11110100101100010010000001010101111010011000001100100010110101101011011011100101110100',
                                '1972-02-29':::DATE,
                                '20:46:11.595514':::TIME,
                                '1970-01-08 01:12:11.000277+00:00':::TIMESTAMP,
                                (-8.163341714805918362E+30):::DECIMAL,
                                '38a690f6-5b84-455e-a176-0016f5867c53':::UUID,
                                657857402:::OID,
                                0.6257669233131637:::FLOAT8,
                                NULL
                            ),
                            (
                                '3c13:2ef:246f:f3f6:30e4:4252:b3f7:cd6/60':::INET,
                                true,
                                B'001011000',
                                NULL,
                                'bc057ccb-72b5-4259-a429-7f9acef966c3':::UUID,
                                4185751985:::OID,
                                '1970-01-07 20:15:16.000086+00:00':::TIMESTAMPTZ,
                                e'\\\\xdc':::BYTES,
                                B'111110011100100100100010101',
                                '1973-04-03':::DATE,
                                '07:55:13.12306':::TIME,
                                '1970-01-11 09:05:30.000905+00:00':::TIMESTAMP,
                                2.528568169792044095E+19:::DECIMAL,
                                '8163e5b8-0622-48c6-9007-0cd104d01a9e':::UUID,
                                NULL,
                                (-1.3052014485532828):::FLOAT8,
                                '14:10:50.700953':::TIME
                            )
                    )
                        AS tab_110
                            (
                                col_535,
                                col_536,
                                col_537,
                                col_538,
                                col_539,
                                col_540,
                                col_541,
                                col_542,
                                col_543,
                                col_544,
                                col_545,
                                col_546,
                                col_547,
                                col_548,
                                col_549,
                                col_550,
                                col_551
                            )
        ),
    with_47 (col_552)
        AS (
            SELECT
                *
            FROM
                (VALUES (NULL), (5668680070121424285:::INT8), ((-1521135999116229970):::INT8), ((-563068193391728125):::INT8), (7538746715150594749:::INT8))
                    AS tab_111 (col_552)
        )
SELECT
    with_46.col_484 AS col_553
FROM
    with_46
WHERE
    with_46.col_485
ORDER BY
    with_46.col_498 DESC
LIMIT
    9:::INT8;
```

```
panic: rowCount passed in was too small [recovered]
        panic: rowCount passed in was too small [recovered]
        panic: panic while executing 1 statements: WITH _ (_, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _) AS (SELECT * FROM (VALUES (CASE WHEN CASE WHEN _ THEN _ ELSE _ END THEN _ ELSE _ - _ END | CASE WHEN not_like_escape(_, _, _) THEN COALESCE(_:::INET, inet_server_port() + _) ELSE _:::INET END, _, _, _:::INT8, _:::UUID, _:::OID, _:::TIMESTAMPTZ, _, _, _:::DATE, _:::TIME, _:::TIMESTAMP, _:::DECIMAL, _, _:::OID, _:::FLOAT8, _)) AS _ (_, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _) EXCEPT SELECT * FROM (VALUES (_:::INET, _, _, (_):::INT8, _:::UUID, _, _, _:::BYTES, _, _:::DATE, _:::TIME, _:::TIMESTAMP, _:::DECIMAL, _:::UUID, _:::OID, _:::FLOAT8, _:::TIME)) AS _ (_, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _) INTERSECT SELECT * FROM (VALUES (_:::INET, _, _, _:::INT8, _:::UUID, _:::OID, _:::TIMESTAMPTZ, _, _, _:::DATE, _, _:::TIMESTAMP, _:::DECIMAL, _:::UUID, _:::OID, _:::FLOAT8, _:::TIME)) AS _ (_, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _) EXCEPT SELECT * FROM (VALUES (_:::INET, _, _, _:::INT8, _:::UUID, _:::OID, _:::TIMESTAMPTZ, _:::BYTES, _, _:::DATE, _, _:::TIMESTAMP, _:::DECIMAL, _:::UUID, _:::OID, _:::FLOAT8, _), (__more2__)) AS _ (_, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _, _)) , _ (_) AS (SELECT * FROM (VALUES (_), (__more4__)) AS _ (_)) SELECT _._ AS _ FROM _ WHERE _._ ORDER BY _._ DESC LIMIT _:::INT8; caused by rowCount passed in was too small

goroutine 506 [running]:
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0xc0015fa000, 0x38ef9c0, 0xc001719400, 0x2c90820, 0x38a8fc0)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:715 +0x36d
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1(0xc0015fa000, 0x38ef9c0, 0xc001719400)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:427 +0x61
panic(0x2c90820, 0x38a8fc0)
        /usr/local/go/src/runtime/panic.go:513 +0x1b9
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder.(*Builder).Build.func1(0xc00158c858)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/builder.go:142 +0x9a
panic(0x2c90820, 0x38a8fc0)
        /usr/local/go/src/runtime/panic.go:513 +0x1b9
github.com/cockroachdb/cockroach/pkg/sql/opt/memo.(*statisticsBuilder).selectivityFromNullCounts(0xc0016a0670, 0x4, 0x0, 0x3928bc0, 0xc0005d2e90, 0xc0005d2d98, 0x3ff0000000000000, 0x3ff0000000000000)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/opt/memo/statistics_builder.go:2494 +0x209
github.com/cockroachdb/cockroach/pkg/sql/opt/memo.(*statisticsBuilder).buildSelect(0xc0016a0670, 0xc0005d2e90, 0xc0005d2d28)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/opt/memo/statistics_builder.go:627 +0x4b6
github.com/cockroachdb/cockroach/pkg/sql/opt/memo.(*logicalPropsBuilder).buildSelectProps(0xc0016a0660, 0xc0005d2e90, 0xc0005d2d28)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/opt/memo/logical_props_builder.go:236 +0x3c1
github.com/cockroachdb/cockroach/pkg/sql/opt/memo.(*Memo).MemoizeSelect(0xc0016a0580, 0x3927c20, 0xc000ed3a30, 0xc0014d5170, 0x1, 0x1, 0xc000ad6005, 0x7)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/opt/memo/expr.og.go:12053 +0x14e
github.com/cockroachdb/cockroach/pkg/sql/opt/norm.(*Factory).ConstructSelect(0xc0015fab10, 0x3927c20, 0xc000ed3a30, 0xc0014d5170, 0x1, 0x1, 0x0, 0x3911d80)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/opt/norm/factory.og.go:750 +0x537b
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder.(*Builder).buildWhere(0xc00143e000, 0xc000368b40, 0xc0016891e0)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/select.go:723 +0x246
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder.(*Builder).buildSelectClause(0xc00143e000, 0xc0014d4750, 0xc000958098, 0x1, 0x1, 0x0, 0x0, 0x0, 0xc0016880f0, 0x203000)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/select.go:631 +0x8f
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder.(*Builder).buildSelect(0xc00143e000, 0xc0017a5b00, 0x0, 0x0, 0x0, 0xc0016880f0, 0x0)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/select.go:585 +0x201
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder.(*Builder).buildStmt(0xc00143e000, 0x38f55c0, 0xc0017a5b00, 0xc001688000, 0x4)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/builder.go:219 +0x10e
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder.(*Builder).Build(0xc00143e000, 0x0, 0x0)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/builder.go:156 +0x111
github.com/cockroachdb/cockroach/pkg/sql.(*optPlanningCtx).buildExecMemo(0xc0015faab0, 0x38efa80, 0xc001817620, 0x0, 0xde9843, 0x38efa80, 0xc001817620)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/plan_opt.go:421 +0x18e
github.com/cockroachdb/cockroach/pkg/sql.(*planner).makeOptimizerPlan(0xc0015fa350, 0x38efa80, 0xc001817620, 0x31af956, 0x19, 0x178ead85e73f1, 0x382affc9)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/plan_opt.go:155 +0xa2
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).makeExecPlan(0xc0015fa000, 0x38efa80, 0xc001817620, 0xc0015fa350, 0x6, 0x6bbc0f)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:979 +0x131
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).dispatchToExecutionEngine(0xc0015fa000, 0x38efa80, 0xc001817620, 0xc0015fa350, 0x7fe929f2cd78, 0xc0010562c0, 0x0, 0x0)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:866 +0x148
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmtInOpenState(0xc0015fa000, 0x38efa80, 0xc001817620, 0x38f55c0, 0xc0017a5b00, 0xc000ad6000, 0x126f, 0x0, 0x0, 0x0, ...)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:457 +0xdd4
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).execStmt(0xc0015fa000, 0x38efa80, 0xc001817620, 0x38f55c0, 0xc0017a5b00, 0xc000ad6000, 0x126f, 0x0, 0x0, 0x0, ...)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:103 +0x610
github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).run(0xc0015fa000, 0x38ef9c0, 0xc001719400, 0xc00004de38, 0x5400, 0x15000, 0xc00004ded0, 0xc0003b2b20, 0x0, 0x0)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1178 +0x21bb
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn(0xc00080d520, 0x38ef9c0, 0xc001719400, 0xc0015fa000, 0x5400, 0x15000, 0xc00004ded0, 0xc0003b2b20, 0x0, 0x0)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:429 +0xce
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).serveImpl.func4(0xc00080d520, 0xc00034bb00, 0xc000398494, 0xc0003b2b20, 0x38ef9c0, 0xc001719400, 0xc0015fa000, 0x5400, 0x15000, 0xc00004ded0, ...)
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:332 +0xcf
created by github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*conn).serveImpl
        /home/mjibson/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:315 +0x1040
```

Found with sqlsmith.

--- a/pkg/sql/opt/memo/statistics_builder.go
+++ b/pkg/sql/opt/memo/statistics_builder.go
@@ -1452,11 +1452,12 @@ func (sb *statisticsBuilder) buildSetNode(setNode RelExpr, relProps *props.Relat
 	switch setNode.Op() {
 	case opt.UnionOp, opt.IntersectOp, opt.ExceptOp:
 		// Since UNION, INTERSECT and EXCEPT eliminate duplicate rows, the row
-		// count will equal the distinct count of the set of output columns.
+		// count will equal the distinct count of the set of output columns, plus
+		// any null values.
 		setPrivate := setNode.Private().(*SetPrivate)
 		outputCols := setPrivate.OutCols.ToSet()
 		colStat := sb.colStatSetNodeImpl(outputCols, setNode, relProps)
-		s.RowCount = colStat.DistinctCount
+		s.RowCount = min(colStat.DistinctCount+colStat.NullCount, s.RowCount)
 	}
 
 	sb.finalizeFromCardinality(relProps)
@@ -1558,9 +1559,11 @@ func (sb *statisticsBuilder) colStatValues(
 		hash := uint64(offset64)
 		hasNull := false
 		for i, elem := range row.(*TupleExpr).Elems {
-			if elem.Op() == opt.NullOp {
-				hasNull = true
-			} else if colSet.Contains(int(values.Cols[i])) {
+			if colSet.Contains(int(values.Cols[i])) {
+				if elem.Op() == opt.NullOp {
+					hasNull = true
+					break
+				}
 				// Use the pointer value of the scalar expression, since it's already
 				// been interned. Therefore, two expressions with the same pointer
 				// have the same value.
@@ -1569,11 +1572,10 @@ func (sb *statisticsBuilder) colStatValues(
 				hash *= prime64
 			}
 		}
-		if hash != offset64 {
-			distinct[hash] = struct{}{}
-		}
 		if hasNull {
 			nullCount++
+		} else {
+			distinct[hash] = struct{}{}
 		}
 	}
 

--- a/pkg/sql/opt/memo/testdata/memo
+++ b/pkg/sql/opt/memo/testdata/memo
@@ -274,7 +274,7 @@ memo (optimized, ~4KB, required=[presentation: x:7,y:8])
  ├── G1: (union G2 G3)
  │    └── [presentation: x:7,y:8]
  │         ├── best: (union G2 G3)
- │         └── cost: 2149.84
+ │         └── cost: 2150.04
  ├── G2: (scan a)
  │    └── []
  │         ├── best: (scan a)

--- a/pkg/sql/opt/memo/testdata/stats/set
+++ b/pkg/sql/opt/memo/testdata/stats/set
@@ -712,7 +712,7 @@ union
  ├── columns: x:9(int)
  ├── left columns: b.x:1(int)
  ├── right columns: c.x:5(int)
- ├── stats: [rows=10000, distinct(9)=10000, null(9)=2]
+ ├── stats: [rows=10002, distinct(9)=10000, null(9)=2]
  ├── key: (9)
  ├── project
  │    ├── columns: b.x:1(int)
@@ -738,7 +738,7 @@ intersect
  ├── columns: x:1(int)
  ├── left columns: b.x:1(int)
  ├── right columns: c.x:5(int)
- ├── stats: [rows=5000, distinct(1)=5000, null(1)=1]
+ ├── stats: [rows=5001, distinct(1)=5000, null(1)=1]
  ├── key: (1)
  ├── project
  │    ├── columns: b.x:1(int)
@@ -764,7 +764,7 @@ except
  ├── columns: x:1(int)
  ├── left columns: b.x:1(int)
  ├── right columns: c.x:5(int)
- ├── stats: [rows=5000, distinct(1)=5000, null(1)=1]
+ ├── stats: [rows=5001, distinct(1)=5000, null(1)=1]
  ├── key: (1)
  ├── project
  │    ├── columns: b.x:1(int)
@@ -857,3 +857,40 @@ except-all
            ├── stats: [rows=10000, distinct(5)=5000, null(5)=1000, distinct(7)=10, null(7)=7500, distinct(5,7)=10000, null(5,7)=7750]
            ├── key: (8)
            └── fd: (8)-->(5-7)
+
+# Regression test for #35715.
+opt colstat=(5,2)
+SELECT * FROM
+((VALUES (NULL, true) EXCEPT (VALUES (1, NULL)))) AS t(a, b)
+WHERE a IS NULL and b
+----
+select
+ ├── columns: a:5(int) b:2(bool!null)
+ ├── cardinality: [0 - 1]
+ ├── stats: [rows=1, distinct(2)=1, null(2)=0, distinct(5)=0, null(5)=1, distinct(2,5)=0, null(2,5)=1]
+ ├── key: ()
+ ├── fd: ()-->(2,5)
+ ├── except
+ │    ├── columns: column1:5(int) column2:2(bool)
+ │    ├── left columns: column1:5(int) column2:2(bool)
+ │    ├── right columns: column1:3(int) column2:6(bool)
+ │    ├── cardinality: [0 - 1]
+ │    ├── stats: [rows=1, distinct(2)=1, null(2)=0, distinct(5)=0, null(5)=1, distinct(2,5)=0, null(2,5)=1]
+ │    ├── key: (2,5)
+ │    ├── values
+ │    │    ├── columns: column2:2(bool) column1:5(int)
+ │    │    ├── cardinality: [1 - 1]
+ │    │    ├── stats: [rows=1, distinct(2)=1, null(2)=0, distinct(5)=0, null(5)=1, distinct(2,5)=0, null(2,5)=1]
+ │    │    ├── key: ()
+ │    │    ├── fd: ()-->(2,5)
+ │    │    └── (true, NULL) [type=tuple{bool, int}]
+ │    └── values
+ │         ├── columns: column1:3(int) column2:6(bool)
+ │         ├── cardinality: [1 - 1]
+ │         ├── stats: [rows=1, distinct(3)=1, null(3)=0, distinct(6)=0, null(6)=1, distinct(3,6)=0, null(3,6)=1]
+ │         ├── key: ()
+ │         ├── fd: ()-->(3,6)
+ │         └── (1, NULL) [type=tuple{int, bool}]
+ └── filters
+      ├── column1 IS NULL [type=bool, outer=(5), constraints=(/5: [/NULL - /NULL]; tight), fd=()-->(5)]
+      └── variable: column2 [type=bool, outer=(2), constraints=(/2: [/true - /true]; tight), fd=()-->(2)]

--- a/pkg/sql/opt/memo/testdata/stats/values
+++ b/pkg/sql/opt/memo/testdata/stats/values
@@ -97,8 +97,30 @@ SELECT * FROM (VALUES (NULL,1), (1,NULL), (NULL,NULL), (1,2))
 values
  ├── columns: column1:1(int) column2:2(int)
  ├── cardinality: [4 - 4]
- ├── stats: [rows=4, distinct(1,2)=2, null(1,2)=3]
+ ├── stats: [rows=4, distinct(1,2)=1, null(1,2)=3]
  ├── (NULL, 1) [type=tuple{int, int}]
  ├── (1, NULL) [type=tuple{int, int}]
  ├── (NULL, NULL) [type=tuple{int, int}]
  └── (1, 2) [type=tuple{int, int}]
+
+# Regression test for #35715.
+norm colstat=1 colstat=2
+SELECT * FROM (VALUES (NULL, 1))
+----
+values
+ ├── columns: column1:1(unknown) column2:2(int)
+ ├── cardinality: [1 - 1]
+ ├── stats: [rows=1, distinct(1)=0, null(1)=1, distinct(2)=1, null(2)=0]
+ ├── key: ()
+ ├── fd: ()-->(1,2)
+ └── (NULL, 1) [type=tuple{unknown, int}]
+
+norm colstat=1 colstat=2 colstat=(1,2)
+SELECT * FROM (VALUES (NULL, 2), (2, NULL))
+----
+values
+ ├── columns: column1:1(int) column2:2(int)
+ ├── cardinality: [2 - 2]
+ ├── stats: [rows=2, distinct(1)=1, null(1)=1, distinct(2)=1, null(2)=1, distinct(1,2)=0, null(1,2)=2]
+ ├── (NULL, 2) [type=tuple{int, int}]
+ └── (2, NULL) [type=tuple{int, int}]

--- a/pkg/sql/opt/xform/testdata/coster/set
+++ b/pkg/sql/opt/xform/testdata/coster/set
@@ -26,8 +26,8 @@ union
  ├── columns: k:8(int) i:9(int)
  ├── left columns: a.k:1(int) a.i:2(int)
  ├── right columns: x:5(int) z:6(int)
- ├── stats: [rows=1990, distinct(8,9)=1990, null(8,9)=19.92]
- ├── cost: 2149.93
+ ├── stats: [rows=2000, distinct(8,9)=1990, null(8,9)=19.92]
+ ├── cost: 2150.03
  ├── key: (8,9)
  ├── scan a
  │    ├── columns: a.k:1(int!null) a.i:2(int)
@@ -67,8 +67,8 @@ intersect
  ├── columns: k:1(int) i:2(int)
  ├── left columns: k:1(int) i:2(int)
  ├── right columns: x:5(int) z:6(int)
- ├── stats: [rows=990, distinct(1,2)=990, null(1,2)=9.91]
- ├── cost: 2139.93
+ ├── stats: [rows=999.91, distinct(1,2)=990, null(1,2)=9.91]
+ ├── cost: 2140.0291
  ├── key: (1,2)
  ├── scan a
  │    ├── columns: k:1(int!null) i:2(int)
@@ -108,8 +108,8 @@ except
  ├── columns: k:1(int) i:2(int)
  ├── left columns: k:1(int) i:2(int)
  ├── right columns: x:5(int) z:6(int)
- ├── stats: [rows=990, distinct(1,2)=990, null(1,2)=9.91]
- ├── cost: 2139.93
+ ├── stats: [rows=999.91, distinct(1,2)=990, null(1,2)=9.91]
+ ├── cost: 2140.0291
  ├── key: (1,2)
  ├── scan a
  │    ├── columns: k:1(int!null) i:2(int)

